### PR TITLE
Add the `SqliteDosStorage` storage backend 

### PR DESCRIPTION
**Pull request by sphuber**
The implementation subclasses the `PsqlDosBackend` and replaces the
PostgreSQL database with an sqlite database. By doing so, the
initialization of the storage only requires a directory on the local
file system where it will create the sqlite file for the database and a
container for the disk-objectstore.

The advantage of this `sqlite_dos` storage over the default `psql_dos`
is that it doesn't require a system service like PostgreSQL. As a
result, creating the storage is very straightforward and can be done
with almost no setup. The advantage over the existing `sqlite_zip` is
that the `sqlite_dos` is not read-only but can be used to write data
as well.

Combined with the `verdi profile setup` command, a working profile can
be created with a single command:

    verdi profile setup core.sqlite_dos -n --profile sqlite-dos

This makes this storage backend very useful for tutorials and demos
that don't rely on performance.

### CODE REVIEW

**Comment by danielhollas** [comment]
@sphuber this is awesome 👍. 

It would also be perfect if it was possible to use this for the test suite. Perhaps there could be some pytest switch that would allow changing the storage backend? It is also currently a bit annoying that the backend is loaded unconditionally, even if you run just a single test file that does not require it. Again, a pytest switch that disables the session fixture that loads the backend would be awesome (and I suspect would speed up iteration during development).

File: aiida/manage/configuration/config.py
Comment (giovannipizzi) [inline aiida/manage/configuration/config.py:102]: Shall this still be speaking explicitly of PostgreSQL? Maybe OK to keep but add "E.g., to avoid hitting..."

**Review by giovannipizzi — APPROVED** [review]
Indeed, great! I left a small comment for a documentation string (maybe you can check if there are other similar cases in the code).

I fully support Daniel's comment - maybe open an issue and we look into it once this is merged?

I also opened https://github.com/aiidateam/aiida-tutorials/issues/468 for the future, when we'll prepare the next tutorial

Finally, one important note. In the previous coding week, we've been discussing having a documentation page for the various backends, and have it properly linked in the docs so people see it when they setup AiiDA. And this was combined with a simpler verdi profile setup mechanism. I think @superstar54 , @unkcpz and/or @eimrek might have been working on this? Has this ever been merged? I recommend adding the documentation for these new backend as well. Either we open an issue now and plan to fix it soon, or if that documentation is already in the docs, we could fix it directly in this PR

**Comment by sphuber** [comment]
Thanks for the review @giovannipizzi 

> I fully support Daniel's comment - maybe open an issue and we look into it once this is merged?

Regarding switching off the automatic loading of the backend in the session fixture: we had discussed this in the past, but decided against it since it is used by plugins and so changing it would be breaking, as they would have to start adding it manually. I agree that having it off by default would still be better, so maybe we can revisit whether this is an acceptable breaking change for plugins.

As for changing the backend plugin: I already have the required scaffolding in place in the [`aiida-s3`](https://github.com/sphuber/aiida-s3/blob/master/tests/conftest.py) plugin. We can repurpose this and add it to `aiida-core`.

> I also opened [aiidateam/aiida-tutorials#468](https://github.com/aiidateam/aiida-tutorials/issues/468) for the future, when we'll prepare the next tutorial
> Finally, one important note. In the previous coding week, we've been discussing having a documentation page for the various backends, and have it properly linked in the docs so people see it when they setup AiiDA. And this was combined with a simpler verdi profile setup mechanism. I think @superstar54 , @unkcpz and/or @eimrek might have been working on this? Has this ever been merged? I recommend adding the documentation for these new backend as well. Either we open an issue now and plan to fix it soon, or if that documentation is already in the docs, we could fix it directly in this PR

I was not aware of this effort but agree that this would be very useful. This ties neatly into why we cannot merge this PR just yet. There is still some functionality that is not supported by this backend, mostly related to the querybuilder:
* use of `date_trunc` in `aiida.storage.psql_dos.orm.querybuilder.QueryBuilder.get_creation_statistics`
* queries that use JSONB syntax
* `Group.count()`
If I cannot fix these, then we should clearly document this at the bare minimum, _and_ we should stress that this backend should really only be used for testing, demoing and other non-production type activities.

Since this PR was anyway dependent on a bunch of other PRs being merged first, I had put these final details on hold. But will try to find time soon to work on them.

**Comment by unkcpz** [comment]
>  I think @superstar54 , @unkcpz and/or @eimrek might have been working on this? Has this ever been merged? 

@giovannipizzi I missed this comment, sorry. 
I think you mentioned the https://github.com/aiidateam/aiida-core/pull/6070 draft done by @superstar54 during coding week if I remember correctly. Not sure there are overlap work but it was mentioned the `Sqlite` backend is in aiida-core already.

**Comment by sphuber** [comment]
> Not sure there are overlap work but it was mentioned the `Sqlite` backend is in aiida-core already.

The PR of @superstar54 uses the `SqliteTempBackend` which already exists in `aiida-core`. Here I am adding _another_ backend based on `sqlite` though, the `SqliteDosStorage`. The difference is that the latter is persistent, whereas the `SqliteTempBackend` gets destroyed as soon as the Python interpreter shuts down.

**Comment by sphuber** [comment]
This is blocked by #6173 Currently it is necessary to specify `--first-name`, `--last-name` and `--institution` with `verdi profile setup -n` making it way less easy to create a new profile. This should be fixed first.

**Pull request merged**
